hermes/hermes.go

```diff
@@ -0,0 +1,95 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// hermes.go defines the structures necessary for the main hermes object to
+// monitor a storage system.
+
+package hermes
+
+import (
+	"github.com/google/cloudprober"
+)
+
+// FileOperation is an int used as part of the FileOperation enum within the intent log of Hermes' StateJournal.
+type FileOperation int
+
+// FileOperation enum is used for marking the operation intent within the StateJournal of Hermes.
+// FileOperation has two possible values: CREATE, DELETE.
+const (
+	Create FileOperation = iota
+	Delete
+)
+
+// String() allows the FileOperation constants to be conveniently converted to a print-friendly format.
+// Returns:
+// - string: Returns the print-friendly string version of the fileOperation enum.
+func (fileOperation FileOperation) String() string {
+	return [...]string{"Create", "Delete"}[fileOperation]
+}
+
+// Intent stores the next intended file operation of Hermes.
+// This is used as part of the StateJournal of Hermes.
+type Intent struct {
+	operation FileOperation `json:"fileoperation"` // Stores the file operation intent, either CREATE or DELETE
+	filename  string        `json:"filename"`      // Stores the filename that the operation is being performed on.
+}
+
+// StateJournal stores the state of Hermes in two parts: the next operation intent and a map of filenames.
+// The intent stores the operation to be performed and the name of the file that the operation is being performed on.
+// The filenames map is a map of file IDs to filenames.
+// If an entry does not exist for a given ID, then the file does not exist (i.e. has been deleted)
+type StateJournal struct {
+	// intent stores the next intended file operation and the name of the file that the operation is being performed on.
+	intent Intent `json:"intent"`
+	// filenames is a map of file IDs to filenames.
+	filenames map[int]string `json:"filenames"`
+}
+
+// Init initialises the map in the StateJournal so that entries can be added to it.
+func (sj *StateJournal) Init() {
+	sj.filenames = make(map[int]string) // initialise filenames to a map
+}
+
+// Hermes is the main Hermes prober that will startup Hermes and initiate monitoring targets.
+type Hermes struct {
+	// Journal stores the state of Hermes as a combination of next operation intent and a filenames map
+	Journal StateJournal
+	// CancelCloudprober is a cancel() function associated with the context passed to Cloudprober when initialised.
```

Comment:
When is the user supposed to call it?
What happens when it's called?
Can it be called muliple times?
Does it *have* to be called? What happens if it's never called?

Reply:
My current expectation is that this function shouldn't be called by the user of Hermes (the project), but instead will be called to gracefully shutdown Hermes. For example, we could have a `StopHermes()` method that cancels this context to cancel Cloudprober and shuts down Hermes after a configurable period of time.

When called, it cancels the context passed to Cloudprober which stops any goroutines and shuts down Cloudprober. 

It is possible to call multiple times, but it wouldn't make sense to, as it's ineffectual after the first call.

It doesn't have to be called. If it's never called, then Cloudprober won't shutdown unless Hermes itself is shutdown, for example using SIGINT or SIGTERM.

Please let me know if you disagree with any of these points.

I will add them to the comment above this field.

Example:
https://github.com/google/cloudprober/blob/4cbdccb155c095efd4a7ace735fc0bd26c8c462d/cmd/cloudprober.go#L192-L198

Reply:
I would remove this for now and only add it when it's actually required and called.
My questions where rhetorical: these are questions that should be answered by the docstring.

Reply:
Okay, I'm happy to remove it for now. I thought the questions were rhetorical and for the info to be added to the docstring, but I'm also new to contexts and wasn't sure if my understanding was fully correct.